Unable to replace a lost node without downtime

I'm trying, on a testing minio cluster with 4 nodes (1 disk each) to drop a node, replace it, and recover to nominal state without downtime.
I can't achieve this, either due to a bug or a lack of documentation.

## Expected Behavior
node fails and gets replaced.
Run Heal on new node which replaces the failed node
Whole cluster (each node) becomes stable in nominal state

## Current Behavior
node fails ang gets replaced.
Run Heal on new node which replaces the failed node
Replaced node health is ok but all other nodes can't connect to replaced node anymore and cluster stay unhealthy.
Stopping and restarting all nodes (inducing a downtime) let the cluster come back to a stable state

## Possible Solution
no re-connect after disk failure ?

## Steps to Reproduce (for bugs)
See details below
1. Build a 4-nodes cluster (with 1 disk each)
2. simulate node replacement: Stop minio on a node and delete all data on the disk (including `.minio.sys`), then sart minio again
3. mc admin heal on replaced server => `.minio.sys` gets rebuilded
4. mc admin heal -r on replaced server => data on this node gets rebuilded
5. all 3 other nodes can't connect to newly replaced disk with traceback in minio logs

```
[ERROR] Unable to fetch disk info for &cmd.networkStorage{rpcClient:(*cmd.AuthRPCClient)(0xc420542480), connected:false} (disk not found)
```

## Context
I know minio is primarily designed to be installed and forgottent and failed disks never replaced, but I need it as a transition for legacy applications and I want to be able to recover on failed disks

## Your Environment
* Version used (`minio version`): 
```
# minio version
Version: 2018-03-19T19:22:06Z
Release-Tag: RELEASE.2018-03-19T19-22-06Z
Commit-ID: a4d2e2b4283b228c34ca2fe0aa007d792a726c3a
```
* Server type and version: 9 (stretch) x64

## Details

Having a 4 node cluster with 1 disk each. We call each node minio-0{1...4}

```
minio-03# mc admin service status local
Uptime: 1 minutes 35 seconds.
minio-03# mc admin info local
●  localhost:9000
   Uptime : online since 1 minute ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 170KiB, Outgoing 64KiB
  Storage : Total 11GiB, Free 11GiB
    Disks : 4, 0


●  10.230.0.21:9000
   Uptime : online since 1 minute ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 157KiB, Outgoing 67KiB
  Storage : Total 11GiB, Free 11GiB
    Disks : 4, 0


●  10.230.0.22:9000
   Uptime : online since 1 minute ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 160KiB, Outgoing 65KiB
  Storage : Total 11GiB, Free 11GiB
    Disks : 4, 0


●  10.230.0.24:9000
   Uptime : online since 1 minute ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 170KiB, Outgoing 66KiB
  Storage : Total 11GiB, Free 11GiB
    Disks : 4, 0


minio-03# mc admin heal -n -r minio-01/
 ◑  test2
    0/3 objects; 402 KiB in 2s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 7 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
minio-03# mc admin heal -n -r minio-02/
 ◓  test/fic2
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 7 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
minio-03# mc admin heal -n -r minio-03/
 ◓  mybucket/hosts
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 7 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
minio-03# mc admin heal -n -r minio-04/
 ◓  test/fic2
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 7 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```

### STOP minio on minio-03 and delete everything (including .minio.sys) on minio disk (same as node rebuild)

```
minio-03# mc admin info minio-01
●  localhost:9000
   Uptime : online since 18 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 258KiB, Outgoing 110KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.22:9000
   Uptime : online since 18 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 264KiB, Outgoing 109KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.23:9000
   Uptime : Server is offline
    Error : dial-http 10.230.0.23:9000/minio/admin: dial tcp 10.230.0.23:9000: getsockopt: connection refused

●  10.230.0.24:9000
   Uptime : online since 18 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 271KiB, Outgoing 111KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1
```

### START minio again on minio-03 (i.e. rebuild a fresh replacement node)

```
minio-03# mc admin info local
●  localhost:9000
   Uptime : online since 28 seconds ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 20KiB, Outgoing 13KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.21:9000
   Uptime : online since 29 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 284KiB, Outgoing 129KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.22:9000
   Uptime : online since 29 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 290KiB, Outgoing 120KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.24:9000
   Uptime : online since 29 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 298KiB, Outgoing 122KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1

minio-03# mc admin heal minio-03
 ◓  mybucket
    0/0 objects; 0 B in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 1 │  25.0% ███          │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 3 │  75.0% █████████    │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```
=> .minio.sys gets rebuilt but no data rebuilt

```
minio-03# mc admin heal -r minio-03/
 ◓  mybucket/hosts
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 4 │  57.1% ███████      │
    │ Yellow │ 3 │  42.9% ██████       │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```
Every data gets back ok on minio-03:
```
# mc admin heal -r -n minio-03/
 ◓  test/fic2
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 7 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```

BUT minio-01, 02 and 04 can't see minio-03 anymore and can't get back to an "healthy" state :
```
# mc admin info minio-03
●  localhost:9000
   Uptime : online since 4 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 118KiB, Outgoing 84KiB
  Storage : Total 11GiB, Free 11GiB
    Disks : 4, 0


●  10.230.0.21:9000
   Uptime : online since 33 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 485KiB, Outgoing 270KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.22:9000
   Uptime : online since 33 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 488KiB, Outgoing 255KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1


●  10.230.0.24:9000
   Uptime : online since 33 minutes ago
  Version : 2018-03-19T19:22:06Z
   Region : 
 SQS ARNs : <none>
    Stats : Incoming 596KiB, Outgoing 357KiB
  Storage : Total 5.6GiB, Free 5.5GiB
    Disks : 3, 1

# mc admin heal -r -n minio-01
 ◓  test/fic2
    0/3 objects; 402 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 1 │  14.3% ██           │
    │ Yellow │ 3 │  42.9% ██████       │
    │ Red    │ 3 │  42.9% ██████       │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```
and I find in the logs of minio-01 :
```
2018-03-23T14:24:33.067445+01:00 minio-01 minio[25320]: Trace: 1: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:150:cmd.getDisksInfo()
2018-03-23T14:24:33.068655+01:00 minio-01 minio[25320]:        2: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:181:cmd.getStorageInfo()
2018-03-23T14:24:33.069838+01:00 minio-01 minio[25320]:        3: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:220:cmd.xlObjects.StorageInfo()
2018-03-23T14:24:33.070764+01:00 minio-01 minio[25320]:        4: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-sets.go:249:cmd.(*xlSets).StorageInfo()
2018-03-23T14:24:33.071834+01:00 minio-01 minio[25320]:        5: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/admin-handlers.go:522:cmd.adminAPIHandlers.HealHandler()
2018-03-23T14:24:33.073099+01:00 minio-01 minio[25320]:        6: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/admin-router.go:66:cmd.(adminAPIHandlers).HealHandler-fm()
2018-03-23T14:24:33.074087+01:00 minio-01 minio[25320]:        7: /opt/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP()
2018-03-23T14:24:33.075402+01:00 minio-01 minio[25320]:        8: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP()
2018-03-23T14:24:33.076553+01:00 minio-01 minio[25320]:        9: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:617:cmd.rateLimit.ServeHTTP()
2018-03-23T14:24:33.077791+01:00 minio-01 minio[25320]:       10: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:575:cmd.pathValidityHandler.ServeHTTP()
2018-03-23T14:24:33.079006+01:00 minio-01 minio[25320]:       11: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:514:cmd.httpStatsHandler.ServeHTTP()
2018-03-23T14:24:33.081396+01:00 minio-01 minio[25320]:       12: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:65:cmd.requestSizeLimitHandler.ServeHTTP()
2018-03-23T14:24:33.083241+01:00 minio-01 minio[25320]:       13: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:90:cmd.requestHeaderSizeLimitHandler.ServeHTTP()
2018-03-23T14:24:33.084160+01:00 minio-01 minio[25320]:       14: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP()
2018-03-23T14:24:33.085054+01:00 minio-01 minio[25320]:       15: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:231:cmd.redirectHandler.ServeHTTP()
2018-03-23T14:24:33.086165+01:00 minio-01 minio[25320]:       16: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:289:cmd.minioReservedBucketHandler.ServeHTTP()
2018-03-23T14:24:33.087103+01:00 minio-01 minio[25320]:       17: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:258:cmd.cacheControlHandler.ServeHTTP()
2018-03-23T14:24:33.088167+01:00 minio-01 minio[25320]:       18: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:359:cmd.timeValidityHandler.ServeHTTP()
2018-03-23T14:24:33.088971+01:00 minio-01 minio[25320]:       19: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1()
2018-03-23T14:24:33.089440+01:00 minio-01 minio[25320]:       20: /opt/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP()
2018-03-23T14:24:33.089932+01:00 minio-01 minio[25320]:       21: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:465:cmd.resourceHandler.ServeHTTP()
2018-03-23T14:24:33.090404+01:00 minio-01 minio[25320]:       22: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/auth-handler.go:257:cmd.authHandler.ServeHTTP()
2018-03-23T14:24:33.090872+01:00 minio-01 minio[25320]:       23: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:133:cmd.reservedMetadataHandler.ServeHTTP()
2018-03-23T14:24:33.091361+01:00 minio-01 minio[25320]:       24: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/http/server.go:111:http.(*Server).Start.func1()
2018-03-23T14:24:33.091831+01:00 minio-01 minio[25320]:       25: /opt/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP()
2018-03-23T14:24:33.092299+01:00 minio-01 minio[25320]:       26: /opt/go/src/net/http/server.go:2619:http.serverHandler.ServeHTTP()
2018-03-23T14:24:33.092765+01:00 minio-01 minio[25320]:       27: /opt/go/src/net/http/server.go:1801:http.(*conn).serve()
2018-03-23T14:24:33.093259+01:00 minio-01 minio[25320]: [2018-03-23T13:24:33.067030844Z] [ERROR] Unable to fetch disk info for &cmd.networkStorage{rpcClient:(*cmd.AuthRPCClient)(0xc420542480), connected:false} (disk not found)
```

I also see this log entries on node minio-01 when issuing `mc admin info minio-01`:
```
2018-03-23T15:42:23.690175+01:00 minio-01 minio[25320]: Trace: 1: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:150:cmd.getDisksInfo()
2018-03-23T15:42:23.691898+01:00 minio-01 minio[25320]:        2: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:181:cmd.getStorageInfo()
2018-03-23T15:42:23.693494+01:00 minio-01 minio[25320]:        3: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-v1.go:220:cmd.xlObjects.StorageInfo()
2018-03-23T15:42:23.694543+01:00 minio-01 minio[25320]:        4: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/xl-sets.go:249:cmd.(*xlSets).StorageInfo()
2018-03-23T15:42:23.695027+01:00 minio-01 minio[25320]:        5: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/admin-rpc-client.go:141:cmd.localAdminClient.ServerInfoData()
2018-03-23T15:42:23.695530+01:00 minio-01 minio[25320]:        6: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/admin-handlers.go:244:cmd.adminAPIHandlers.ServerInfoHandler.func1()
2018-03-23T15:42:23.696168+01:00 minio-01 minio[25320]: [2018-03-23T14:42:23.689814013Z] [ERROR] Unable to fetch disk info for &cmd.networkStorage{rpcClient:(*cmd.AuthRPCClient)(0xc420542480), connected:false} (disk not found)
```

--- a/cmd/admin-heal-ops.go
+++ b/cmd/admin-heal-ops.go
@@ -217,9 +217,9 @@ func (ahs *allHealState) LaunchNewHealSequence(h *healSequence) (
 	}()
 
 	b, err := json.Marshal(madmin.HealStartSuccess{
-		h.clientToken,
-		h.clientAddress,
-		h.startTime,
+		ClientToken:   h.clientToken,
+		ClientAddress: h.clientAddress,
+		StartTime:     h.startTime,
 	})
 	if err != nil {
 		errorIf(err, "Failed to marshal heal result into json.")

--- a/cmd/xl-sets.go
+++ b/cmd/xl-sets.go
@@ -1105,7 +1105,7 @@ func (s *xlSets) HealBucket(ctx context.Context, bucket string, dryRun bool) (re
 
 	// Check if we had quorum to write, if not return an appropriate error.
 	_, afterDriveOnline := res.GetOnlineCounts()
-	if afterDriveOnline < s.setCount*s.drivesPerSet/2+1 {
+	if afterDriveOnline < ((s.setCount*s.drivesPerSet)/2)+1 {
 		return nil, toObjectErr(errXLWriteQuorum, bucket)
 	}
 

--- a/cmd/xl-v1-healing.go
+++ b/cmd/xl-v1-healing.go
@@ -49,8 +49,8 @@ func checkBucketExistsInQuorum(storageDisks []StorageAPI, bucketName string) (er
 	}
 	wg.Wait()
 
-	writeQuorum := len(storageDisks)/2 + 1
-	return reduceWriteQuorumErrs(errs, nil, writeQuorum)
+	readQuorum := len(storageDisks) / 2
+	return reduceWriteQuorumErrs(errs, nil, readQuorum)
 }
 
 // Heals a bucket if it doesn't exist on one of the disks, additionally

